src/error.rs

```diff
@@ -25,76 +25,191 @@ macro_rules! acheck {
+        return Err(invalid_str(func));
+    };
     let c = unsafe { CStr::from_ptr(s) };
-    let ss = str::from_utf8(c.to_bytes()).map_err(|_| {
-        unsafe { free(s as *mut c_void); }
-        invalid_str(func)
-    })?.to_string();
-    unsafe { free(s as *mut c_void); }
+    let ss = str::from_utf8(c.to_bytes())
+        .map_err(|_| {
+            unsafe {
+                free(s as *mut c_void);
+            }
+            invalid_str(func)
+        })?
+        .to_string();
+    unsafe {
+        free(s as *mut c_void);
+    }
     Ok(ss)
 }
 
 pub fn from_code(func: &'static str, r: c_int) -> Result<c_int> {
-    if r < 0 { Err(Error::new(func, r)) }
-    else { Ok(r) }
+    if r < 0 {
+        Err(Error::new(func, r))
+    } else {
+        Ok(r)
+    }
 }
 
 impl Error {
     pub fn new(func: &'static str, res: c_int) -> Error {
-        let errno = nix::errno::Errno::from_i32(res as i32);
-        Error(func, errno)
+        Self(func, res)
+    }
+
+    pub fn last(func: &'static str) -> Error {
+        Self(
+            func,
+            std::io::Error::last_os_error()
+                .raw_os_error()
+                .unwrap_or_default(),
+        )
     }
 
     pub fn unsupported(func: &'static str) -> Error {
-        Error(func, nix::Error::ENOTSUP)
+        Error(func, libc::ENOTSUP)
     }
 
     /// The function which failed.
-    pub fn func(&self) -> &'static str { self.0 }
-
-
-    /// Underlying error
-    ///
-    /// Match this against the re-export of `nix::Error` in this crate, not against a specific version
-    /// of the nix crate. The nix crate version might be updated with minor updates of this library.
-    pub fn errno(&self) -> nix::Error { self.1 }
+    pub fn func(&self) -> &'static str {
+        self.0
+    }
 
     /// Underlying error
     ///
     /// Match this against the re-export of `nix::Error` in this crate, not against a specific version
     /// of the nix crate. The nix crate version might be updated with minor updates of this library.
-    pub fn nix_error(&self) -> nix::Error { self.1 }
+    pub fn errno(&self) -> i32 {
+        self.1
+    }
 }
 
-pub fn invalid_str(func: &'static str) -> Error { Error(func, nix::Error::EILSEQ) }
+pub fn invalid_str(func: &'static str) -> Error {
+    Error(func, libc::EILSEQ)
+}
 
 impl StdError for Error {
-    fn source(&self) -> Option<&(dyn StdError + 'static)> { Some(&self.1) }
-    fn description(&self) -> &str { "ALSA error" }
+    fn description(&self) -> &str {
+        "ALSA error"
+    }
+}
+
+impl fmt::Debug for Error {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        write!(f, "{}", self)
+    }
 }
 
 impl fmt::Display for Error {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "ALSA function '{}' failed with error '{}'", self.0, self.1)
+        write!(
+            f,
+            "ALSA function '{}' failed with error '{}'",
+            self.0,
+            desc(self.1)
```

Comment:
Would probably make sense to write the number as well as the name here, so you don't end up with 'unknown errno' and no way to troubleshoot further.